ethers-ens/src/main/kotlin/io/ethers/EnsResolver.kt

```diff
@@ -0,0 +1,159 @@
+package io.ethers
+
+import io.ethers.EnsResolver.Error.FailedToResolve
+import io.ethers.EnsResolver.Error.UnknownResolver
+import io.ethers.core.ENSRegistryWithFallback
+import io.ethers.core.FastHex
+import io.ethers.core.PublicResolver
+import io.ethers.core.types.Address
+import io.ethers.core.types.BlockId
+import io.ethers.core.types.Bytes
+import io.ethers.providers.Provider
+import io.ethers.providers.types.RpcResponse
+
+class EnsResolver(private val provider: Provider) {
+
+    /**
+     * Resolve ens name to Address. On error return [RpcResponse] as error.
+     */
+    fun resolveName(ensName: String): RpcResponse<Address> {
+        // Check that ens name is valid
+        if (ensName.isBlank() || (ensName.trim().length == 1 && ensName.contains("."))) {
+            return RpcResponse.error(Error.EnsNameInvalid)
+        }
+
+        val nameHashResponse = NameHash.nameHash(ensName)
+        if (nameHashResponse.isError) return nameHashResponse.propagateError()
+        val nameHash = nameHashResponse.resultOrThrow()
+
+        val resolverResponse = getResolver(nameHash)
+        if (resolverResponse.isError) return resolverResponse.propagateError()
+
+        // Unwrap resolver from RpcResponse and call its addr() function.
+        // If RpcResponse is an error, map it to error FailedToResolve.
+        val resolver = resolverResponse.resultOrThrow()
+        return resolver.addr(Bytes(nameHash))
+            .call(BlockId.LATEST)
+            .sendAwait()
+            .mapError { FailedToResolve(resolver.address, ensName) }
+    }
+
+    /**
+     * Get [PublicResolver] for [nameHash] using [ENSRegistryWithFallback] of current chain.
+     */
+    private fun getResolver(nameHash: ByteArray): RpcResponse<PublicResolver> {
+        return getResolverAddress(nameHash).map { PublicResolver(provider, it) }
+    }
+
+    /**
+     * Get resolver address for [nameHash] from [ENSRegistryWithFallback].
+     */
+    private fun getResolverAddress(nameHash: ByteArray): RpcResponse<Address> {
+        val registryAddr = getRegistryAddress(provider.chainId)
+            ?: return RpcResponse.error(Error.RegistryAddrNotExists(provider.chainId))
+
+        val registryContract = ENSRegistryWithFallback(provider, registryAddr)
+        return registryContract.resolver(Bytes(nameHash))
+            .call(BlockId.LATEST)
+            .map {
+                //
+                if (it == Address.ZERO) {
+                    return@map RpcResponse.error(Error.ResolvingResolver(registryAddr, nameHash.toString()))
+                } else {
+                    return@map RpcResponse.result(it)
+                }
+            }
+            .sendAwait()
+            .mapError { UnknownResolver(registryAddr, FastHex.encodeWithPrefix(nameHash)) }
```

Comment:
`ResolvingResolver` and `UnknownResolver` errors should be switched with one another, however, there is an error when mapping returned error in `map`, and the call returns the error in `mapError`, even when it should return the one in `map`.